File: script/pinglimiter.cs

interval $ping_limiter_tick [

    if (> $playercount 2) [
        
        my cn 0
        my warnings 0
    
        foreach (players) [
            
            set cn $arg1
            
            if (> (player_ping $cn) $ping_limit) [
                
                set warnings (+ 1 (player_var $cn ping_warnings))
                player_var $cn ping_warnings $warnings
                
                if (> $warnings $ping_limit_warnings) [
                    log [Net stats for @(player_name $cn): ping @(player_ping $cn) lag @(player_lag $cn)]
                    kick $cn 0 "server" "high ping"
                    player_var $cn ping_warnings 0
                ] [
                    player_msg $cn [@(green (player_name $cn)), your ping is too high for this server - it needs to be below @ping_limit ms.]
                ]
            
            ]
        ]
        
    ]
    
    if (= $enable_ping_limiter 0) [return -1] //stop interval
]

File: script/pinglimiter.lua
local in_intermission = false
local still_loaded = true

local unloadEventHandlers

local function loadEventHandlers()
    
    local intermission = server.event_handler("intermission", function() in_intermission = true end)
    local mapchange = server.event_handler("mapchange", function() in_intermission = false end)
    
    unloadEventHandlers = function()
    
        server.cancel_handler(intermission)
        server.cancel_handler(mapchange)
        
        still_loaded = false
    end
    
end

server.interval(server.ping_limiter_tick,function()

    if not still_loaded then return -1 end
    if server.playercount < 3 then return end
    
    for player in server.gplayers() do
        
        local warnings = player:pvars().ping_warnings or 0
        
        if (in_intermission == false) and (player:lag() > server.lag_limit) and (player:ping() > server.ping_limit) then
            
            warnings = warnings + 1
            
            if warnings > server.ping_limit_warnings then
                server.log("Net stats for " .. player:name() .. ": ping " .. player:ping() .. " lag " .. player:lag())
                player:kick(1, "server", "high ping")
                player:pvars().ping_warnings = 0
            else
                player:msg(green(player:name()) .. ", your " .. red("ping is too high") .. " for this server - it needs to be " .. orange("below " .. server.ping_limit .. " ms."))
                player:pvars().ping_warnings = warnings
            end
        end
    end
end)

loadEventHandlers()

return {unload = unloadEventHandlers}
